platform/FileHandle.h

```diff
@@ -152,8 +156,88 @@ class FileHandle {
      */
     MBED_DEPRECATED_SINCE("mbed-os-5.4", "Replaced by FileHandle::size")
     virtual off_t flen() { return size(); }
+
+    /** Set blocking or non-blocking mode of the file operation like read/write.
+     *  Definition depends upon the subclass implementing FileHandle.
+     *  The default is blocking.
+     *
+     *  @param blocking true for blocking mode, false for non-blocking mode.
+     */
+    virtual int set_blocking(bool blocking) {
+        return -1;
+    }
+
+    /** Check for poll event flags
+     * The input parameter can be used or ignored - the could always return all events,
+     * or could check just the events listed in events.
+     * Call is non-blocking - returns instantaneous state of events.
+     * Whenever an event occurs, the derived class must call _poll_change().
+     * @param events bitmask of poll events we're interested in - POLLIN/POLLOUT etc.
+     * @return bitmask of poll events that have occurred.
+     */
+    virtual short poll(short events) const {
+        // Possible default for real files
+        return POLLIN | POLLOUT;
+    }
+    /** Returns true if the FileHandle is writable.
+     *  Definition depends upon the subclass implementing FileHandle.
+     *  For example, if the FileHandle is of type Stream, writable() could return
+     *  true when there is ample buffer space available for write() calls.
+     */
+    bool writable() const
+    {
+        return poll(POLLOUT) & POLLOUT;
+    }
+
+    /** Returns true if the FileHandle is readable.
+     *  Definition depends upon the subclass implementing FileHandle.
+     *  For example, if the FileHandle is of type Stream, readable() could return
+     *  true when there is something available to read.
+     */
+    bool readable() const
+    {
+        return poll(POLLIN) & POLLIN;
+    }
+
+    /** Register a callback on state change of the file.
+     *
+     *  The specified callback will be called on state changes such as when
+     *  the file can be written to or read from.
+     *
+     *  The callback may be called in an interrupt context and should not
+     *  perform expensive operations.
+     *
+     *  Note! This is not intended as an attach-like asynchronous api, but rather
+     *  as a building block for constructing  such functionality.
+     *
+     *  The exact timing of when the registered function
+     *  is called is not guaranteed and susceptible to change. It should be used
+     *  as a cue to make read/write/poll calls to find the current state.
+     *
+     *  @param func     Function to call on state change
+     */
+    void sigio(Callback<void()> func);
+
+    /** Issue sigio to user - used by mbed::_poll_change */
+    void _send_sigio()
```

Comment:
@geky - that seems really over-elaborate and error-prone to me. I'm immensely wary of any sort of "cache a value and put it back when we've finished" in a multithreaded environment.  Looking at that I have nightmares of my attempt to unset sigio being unwound as poll exits. **And it still wouldn't work with separate read and write threads both using poll.**

I don't have a problem with removing sigio implementation from FileHandle - once you consider the need to also deal with unblocking read/write, I think it's probably a little neater to put the trio of "unblock us, unblock poll, send sigio" in one place in a wake function, rather than have a sort of two-layer "unblock us, (unblock poll and send sigio)".

But I fundamentally struggling to see the merit of this complexity to avoid a direct FileHandle implementation -> poll() link. What are we trying to achieve here? poll() is part of the "OS" core, in as much as we have one. An implementer of FileHandle has the responsibility to hook into that. There is exactly one poll() implementation.

I can see two benefits of your approach:

1) It avoids linking CondVar and any other support code unnecessarily when no-one is using poll. I can see that's an issue. To cover that simple requirement I'd be inclined to have a single static callback filled in by first use of poll(), so FileHandles call that rather than statically naming poll_change or juggling sigio(). Does make images using poll a tad bigger though.

2) It nearly permits more fine-grained poll implementation - one where by having a specific handle, each FileHandle wakes a specific poll, rather than all polls. But unfortunately it is limited to exactly one poll per FileHandle - regardless of the sigio sharing, you couldn't have two threads for read and write. So it doesn't quite fulfill the promise there.

Systems like Linux require each file handle to have its own wait queue - each poll adds itself to the wait queue of the file handle, which means an arbitrary number of threads can poll and block on each file handle.

All the suggestions so far on this PR are attempts to squeeze sufficient functionality into a non-open-ended system, which may be the problem.

I was aiming to keep this simple, on the assumption that poll() was going to be relatively lightly used compared to sigio and blocking read/writes, and the system is small with few FileHandles, so having every FileHandle wake every poll wouldn't be a big deal.

If we are going to add complexity I'd want to make sure the complexity is worthwhile - I'd want to go far enough to allow that more specific wake-up. Which means per-FileHandle wait lists (which could be CondVars).

Given that precisely NONE of this poll wakeup is being implemented in this PR, I'd be inclined to leave it as future expansion. We can add a new FileHandle API to deal with arbitrary wait queus later, and just make sure that future proper poll() remains compatible with current FileHandles that don't have necessary support - it falls back to timed wait if any specified FileHandles are missing the wakeup support.